include/exec/any_sender_of.hpp

```diff
@@ -0,0 +1,805 @@
+          }
+          void* __p = __get_object_pointer(__self.__storage_);
+          void* __o = __get_object_pointer(__other.__storage_);
+                  // if both object pointers are not null, use the virtual equal_to function
+          return (__p && __o && __get_vtable(__self.__storage_)->__equal_to_(__p, __o)) 
+                  // if both object pointers are nullptrs, they are equal
+                || (!__p && !__o);
+        }
+
+        friend bool operator!=(const __scheduler& __self, const __scheduler& __other) noexcept {
+          return !(__self == __other);
+        }
+
+        __copyable_storage_t<__vtable> __storage_{};
+      };
+
+    template <class _Sigs, class _OldSQs, class _RQ, class... _NewSQs>
+      auto __with_sender_queries_fn(__sender<_Sigs, _OldSQs, _RQ>*, _NewSQs*...) 
+        -> __sender<_Sigs, __minvoke<__mconcat<>, _OldSQs, __types<_NewSQs...>>, _RQ>;
+
+    template <class _Sigs, class _SchQs, class _RQ, class _OldSQs, class... _NewSQs>
+      auto __with_sender_queries_fn(__scheduler<_Sigs, _SchQs, _OldSQs, _RQ>*, _NewSQs*...) 
+        -> __scheduler<_Sigs, _SchQs, __minvoke<__mconcat<>, _OldSQs, __types<_NewSQs...>>, _RQ>;
+
+    template <class _S, class... _Queries>
+      using with_sender_queries = decltype(__with_sender_queries_fn((_S*)nullptr, (_Queries*)nullptr...));
+
+    template <class _Sigs, class _SQs, class _OldRQs, class... _NewRQs>
+      auto __with_receiver_queries_fn(__sender<_Sigs, _SQs, _OldRQs>*, _NewRQs*...) 
+        -> __sender<_Sigs, _SQs, __minvoke<__mconcat<>, _OldRQs, __types<_NewRQs...>>>;
+
+    template <class _S, class... _Queries>
+      using with_receiver_queries = decltype(__with_receiver_queries_fn((_S*)nullptr, (_Queries*)nullptr...));
+
+    template <class _Sigs, class _SchSQs, class _RQ, class _SQs, class... _NewSchSQs>
+      auto __with_scheduler_queries_fn(__scheduler<_Sigs, _SchSQs, _SQs, _RQ>*, _NewSchSQs*...) 
+        -> __scheduler<_Sigs, __minvoke<__mconcat<>, _SchSQs, __types<_NewSchSQs...>>, _SQs, _RQ>;
+
+    template <class _S, class... _Queries>
+      using with_scheduler_queries = decltype(__with_scheduler_queries_fn((_S*)nullptr, (_Queries*)nullptr...));
+
+    // BUGBUG: __completion_signature<_T> returns always true???
+    template <class _Sig>
+      inline constexpr bool __is_compl_sig = false;
+    template <class... _Args>
+      inline constexpr bool __is_compl_sig<set_value_t(_Args...)> = true;
+    template <class _Error>
+      inline constexpr bool __is_compl_sig<set_error_t(_Error)> = true;
+    template <>
+      inline constexpr bool __is_compl_sig<set_stopped_t()> = true;
+
+  template <class... _Sigs>
+    using __sender_of = __t<__any::__sender<completion_signatures<_Sigs...>>>;
+
+  template <class _T>
+    using __transform_value = __if_c<__is_compl_sig<_T>, _T, set_value_t(_T)>;
+
+  } // namepsace __any
+
+  template <class... _Ts>
+    using any_sender_of = __any::__sender_of<__any::__transform_value<_Ts>...>;
```

Comment:
This is clever, but I'm a tad concerned that `any_sender_of<int, short>` might get mistaken for `any_sender_of<set_value_t(int, short)>` instead of `any_sender_of<set_value_t(int), set_value_t(short)>`.

Maybe cap the permitted value types at one. You can additionally require that only the first template parameter is allowed to name a value type; the rest have to be completion signatures.

Reply:
I have no strong feelings for any variation. Is it really too verbose to specify all possible completion signatures? I'm not sure anymore.